### PR TITLE
Issue/192 model property

### PR DESCRIPTION
**Pull request by vallo87**
model property types editable from new index interface
- add, edit, delete
- silent json check for params

Refactored form/page change check: 
- Event delegation instead of direct events
- every elements diff
- elements outside forms can now throw a change event

Minor:
- model index list using flex
- filter label

### CODE REVIEW

**Comment by didoda** [comment]
I tested:
 - `/model/property_types`: add / modify / delete new property types :+1:
 - `/model/property_types`: add property type with unexistant type, do not save it :+1: (ok, but maybe an alert would be a better feedback)
 - `/model/property_types`: add property type with same name of other existent property type, do not save it :+1: (ok, but maybe an alert would be a better feedback here too)
 - `/:object_type/view/:id`:
   - change something and try to go to another page => show alert message :+1:
   - change back (everything like at the beginning) and go to another page => do not show alert message  :+1:

File: src/Controller/ModelController.php [hunk 1]
```diff
@@ -131,4 +133,78 @@ public function view($id) : ?Response
...
+    public function savePropertiesJson()
+    {
+        $payload = $this->request->getData();
+
+        $this->request->allowMethod(['post']);
+        $response = [];
+
+        try {
+            if (!empty($payload)) {
+                extract($payload);
```
Comment (stefanorosanelli) [inline src/Controller/ModelController.php]: I don`t really like `extract()` - code is not so clear and readable, you don't know where variables come from...  we never use it actually

Comment (stefanorosanelli) [reply src/Controller/ModelController.php]: also having 3 methods on add / edit / remove operations will make `savePropertiesJson` method more readable

File: src/Template/Layout/js/app/app.js [hunk 2]
```diff
@@ -374,25 +376,110 @@ const _vueInstance = new Vue({
...
          * @returns {void}
          */
-        alertBeforePageUnload() {
-            var forms = [...document.querySelectorAll('form')];
-            forms.forEach((form) => {
-                form.addEventListener('change', (ev) => {
-                    form.changed = true;
-                });
-                form.addEventListener('submit', (ev) => {
+        alertBeforePageUnload(view) {
```
Comment (stefanorosanelli) [inline src/Template/Layout/js/app/app.js]: do we need this change here? or is it a general optimization?

Probably a PR with new `alert before unload` logic is preferrable, it seems a significant change...

File: src/Template/Layout/js/app/directives/datepicker.js [hunk 3]
```diff
@@ -58,6 +58,8 @@ export default {
 
                 try {
                     let datePicker = flatpickr(element, options);
+                    element.dataset.originalValue = element.value;
+                    // element._flatpickr = datePicker;
```
Comment (stefanorosanelli) [inline src/Template/Layout/js/app/directives/datepicker.js]: remove commented line?

File: config/routes.php [hunk 4]
```diff
@@ -93,6 +93,13 @@
         ['controller' => 'Model', 'action' => 'view'],
         ['pass' => ['id'], '_name' => 'model:view']
     );
+
+    $routes->connect(
+        '/model/:resource_type/savePropertiesJson',
```
Comment (stefanorosanelli) [inline config/routes.php]: can `:resource_type` be other than `property_types` ? 
we could set explicitly ```['resource_types' => 'property_types']```

File: src/Template/Layout/js/app/directives/datepicker.js (hunk 3 above)
Comment (vallo87) [reply src/Template/Layout/js/app/directives/datepicker.js]: that is part of a refactor (did it here cause i needed it in this branch)

File: config/routes.php (hunk 4 above)
Comment (vallo87) [reply config/routes.php]: yes it could used also for object_types and relation

File: src/Controller/ModelController.php (hunk 1 above)
Comment (vallo87) [reply src/Controller/ModelController.php]: ok, i'll refactor it if it's bad practice

File: src/Controller/ModelController.php [hunk 5]
```diff
@@ -131,4 +134,88 @@ public function view($id) : ?Response
...
+            $this->set([
+                'error' => $error->getMessage(),
+                '_serialize' => ['error'],
+            ]);
+
+            return;
+        }
+
+        $this->set((array)$response);
+        $this->set('_serialize', array_keys($response));
```
Comment (stefanorosanelli) [inline src/Controller/ModelController.php]: Use ```$this->set('_serialize',  true)``` to serialize every key

File: src/Controller/ModelController.php [hunk 6]
```diff
@@ -131,4 +134,88 @@ public function view($id) : ?Response
 
         return null;
     }
+
+    /**
+     * save property types (add/edit/delete)
+     *
+     * @return {void}
```
Comment (stefanorosanelli) [inline src/Controller/ModelController.php]: ```@return void``` also to add in  ```public function savePropertiesJson() : void```

File: src/Controller/ModelController.php [hunk 7]
```diff
@@ -131,4 +134,88 @@ public function view($id) : ?Response
...
+     */
+    public function savePropertiesJson()
```
Comment (stefanorosanelli) [inline src/Controller/ModelController.php]: `savePropertyTypesJson` ??

**Comment by didoda** [comment]
A test fails:
```
1) App\Test\TestCase\Controller\AppControllerTest::testBeforeRender
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
-        'jwt' => 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNlcm5hbWUiOiJhZG1pbiIsInJvbGVzIjpbeyJpZCI6MSwibmFtZSI6ImFkbWluIn1dLCJpc3MiOiJodHRwOlwvXC8xMjcuMC4wLjE6ODA4MCIsImlhdCI6MTU0Mzk0MjQ1OSwibmJmIjoxNTQzOTQyNDU5LCJleHAiOjE1NDM5NDM2NTl9.yzQi76o62f6Rw9ZZ_fy-e6S_7bo0dhx4BoEc6t3dxwc'
-        'renew' => 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwOlwvXC8xMjcuMC4wLjE6ODA4MCIsImlhdCI6MTU0Mzk0MjQ1OSwibmJmIjoxNTQzOTQyNDU5LCJzdWIiOjEsImF1ZCI6Imh0dHA6XC9cLzEyNy4wLjAuMTo4MDgwXC9hdXRoIn0.OL6P6J9TOPQ2aluSoG4qWfEq0mnjbiERkuycJGd6qoY'
+        'jwt' => 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNlcm5hbWUiOiJhZG1pbiIsInJvbGVzIjpbeyJpZCI6MSwibmFtZSI6ImFkbWluIn1dLCJpc3MiOiJodHRwOlwvXC8xMjcuMC4wLjE6ODA4MCIsImlhdCI6MTU0Mzk0MjQ1OCwibmJmIjoxNTQzOTQyNDU4LCJleHAiOjE1NDM5NDM2NTh9.Jr087I2Ue6zdxkuWQEiHrsSp_ab88DhfLoKzxzc1wOM'
+        'renew' => 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwOlwvXC8xMjcuMC4wLjE6ODA4MCIsImlhdCI6MTU0Mzk0MjQ1OCwibmJmIjoxNTQzOTQyNDU4LCJzdWIiOjEsImF1ZCI6Imh0dHA6XC9cLzEyNy4wLjAuMTo4MDgwXC9hdXRoIn0.-0dp2F0p7AwhTfFrIwFh2OxTQwrJ8mnu1_gGoyxpNPE'
/home/travis/build/bedita/web/tests/TestCase/Controller/AppControllerTest.php:189
FAILURES!
Tests: 192, Assertions: 335, Failures: 1.
```

[https://travis-ci.org/bedita/web/jobs/463439825]

**Pull request merged**